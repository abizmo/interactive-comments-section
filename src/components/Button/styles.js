import styled from 'styled-components';

const width = {
  big: '9.25rem',
  small: '6.5rem',
};

export const Btn = styled.button`
  align-items: center;
  background: transparent;
  border: none;
  color: ${({ color, theme }) => theme.colors[`${color}-500`]};
  cursor: pointer;
  display: flex;
  font-size: ${({ theme }) => theme.fontSizes[500]};
  font-weight: 500;
  gap: .5rem;
  height: 3rem;
  justify-content: center;
  line-height: 1.1;
  padding: 0;
  transition: opacity 300ms ease-in;

  &:hover {
    opacity: .4;
  }
`;

export const BtnContained = styled(Btn)`
  background: ${({ color, theme }) => theme.colors[`${color}-500`]};
  border-radius: .5rem;
  color: ${({ theme }) => theme.colors['neutral-100']};
  font-size: ${({ theme }) => theme.fontSizes[600]};
  padding: .5rem 1rem;
  text-transform: uppercase;
  width: ${({ size }) => (size !== 'fit' ? width[size] : 'fit-content')};
`;
