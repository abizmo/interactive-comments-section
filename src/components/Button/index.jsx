import React from 'react';
import PropTypes from 'prop-types';
import { Btn, BtnContained } from './styles';

function Button({
  color, icon: Icon, label, onClick, size, variant,
}) {
  let Component = Btn;
  if (variant === 'contained') {
    Component = BtnContained;
  }
  return (
    <Component color={color} onClick={onClick} type="button" size={size}>
      { Icon && (
        <Icon />
      )}
      {label}
    </Component>
  );
}

Button.propTypes = {
  color: PropTypes.oneOf(['neutral', 'primary', 'secondary']),
  icon: PropTypes.elementType,
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  size: PropTypes.oneOf(['fit', 'small', 'big']),
  variant: PropTypes.oneOf(['text', 'contained']),
};

Button.defaultProps = {
  color: 'neutral',
  icon: null,
  size: 'fit',
  variant: 'text',
};

export default Button;
