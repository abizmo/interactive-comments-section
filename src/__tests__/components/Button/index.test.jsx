import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import 'jest-styled-components';

import Button from '../../../components/Button';
import Theme from '../../../theme';

test('renders default button and match snapshot', () => {
  const handleClick = jest.fn();
  const { container } = render(
    <Theme>
      <Button label="Test" onClick={handleClick} />
    </Theme>,
  );

  expect(container).toMatchSnapshot();

  const result = screen.getByText(/Test/);
  expect(result).toHaveStyle('background: transparent');
  expect(result).toHaveStyle('color: hsl( 211, 10%, 45%)');
  expect(result.getAttribute('color')).toEqual('neutral');

  fireEvent.click(result);
  expect(handleClick).toHaveBeenCalled();
});

test('renders primary contained button and match snapshot', () => {
  const { container } = render(
    <Theme>
      <Button
        color="primary"
        label="Test"
        onClick={jest.fn}
        size="big"
        variant="contained"
      />
    </Theme>,
  );

  expect(container).toMatchSnapshot();

  const result = screen.getByText(/Test/);
  expect(result).toHaveStyle('background: hsl(238, 40%, 52%)');
  expect(result).toHaveStyle('color: hsl( 0, 0%, 100%)');
  expect(result).toHaveStyle('width: 8.75rem');
  expect(result.getAttribute('color')).toEqual('primary');
});
