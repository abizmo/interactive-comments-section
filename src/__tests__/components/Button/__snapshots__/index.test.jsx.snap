// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default button and match snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: hsl( 211,10%,45%);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
  font-weight: 500;
  gap: .5rem;
  height: 3rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.1;
  padding: 0;
  -webkit-transition: opacity 300ms ease-in;
  transition: opacity 300ms ease-in;
}

.c0:hover {
  opacity: .4;
}

<div>
  <button
    class="c0"
    type="button"
  >
    Test
  </button>
</div>
`;

exports[`renders primary contained button and match snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: hsl(238,40%,52%);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
  font-weight: 500;
  gap: .5rem;
  height: 3rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.1;
  padding: 0;
  -webkit-transition: opacity 300ms ease-in;
  transition: opacity 300ms ease-in;
}

.c0:hover {
  opacity: .4;
}

.c1 {
  background: hsl(238,40%,52%);
  border-radius: .5rem;
  color: hsl( 0,0%,100%);
  font-size: 1.125rem;
  padding: .5rem 1rem;
  text-transform: uppercase;
  width: 9.25rem;
}

<div>
  <button
    class="c0 c1"
    type="button"
  >
    Test
  </button>
</div>
`;
